File: implementation/01 Fundamentals/Concurreny Basics/Producer-Consumer/java/BoundedBufferTest.java
// BoundedBufferTest.java

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests cover:
 * - Basic correctness: items inserted are items received, in order
 * - Backpressure: producer blocks when buffer is full
 * - Concurrency: no items lost or duplicated under concurrent access
 * - Shutdown: poison pill unblocks waiting consumers
 * - Edge cases: null rejection, zero capacity, single item
 */
class BoundedBufferTest {

    // ---- Basic correctness ----

    @Test
    void singleProducerSingleConsumer_itemsReceivedInOrder() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(10);

        for (int i = 0; i < 5; i++) buffer.put(i);

        for (int i = 0; i < 5; i++) {
            assertEquals(i, buffer.take());
        }
    }

    @Test
    void nullItemRejected() {
        BoundedBuffer<String> buffer = new BoundedBuffer<>(10);
        assertThrows(IllegalArgumentException.class, () -> buffer.put(null));
    }

    @Test
    void zeroCategoryCapacityRejected() {
        assertThrows(IllegalArgumentException.class, () -> new BoundedBuffer<>(0));
        assertThrows(IllegalArgumentException.class, () -> new BoundedBuffer<>(-1));
    }

    // ---- Backpressure ----

    @Test
    @Timeout(3) // must complete within 3 seconds — confirms no infinite block
    void offer_returnsFalse_whenBufferFullAndTimeout() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(2);
        buffer.put(1);
        buffer.put(2); // buffer now full

        // offer with 100ms timeout — should return false quickly, not block forever
        boolean accepted = buffer.offer(3, 100, TimeUnit.MILLISECONDS);
        assertFalse(accepted, "offer should return false when buffer is full");
    }

    @Test
    @Timeout(3)
    void producer_blocksUntilConsumerDrains() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(1);
        buffer.put(1); // fill the buffer

        AtomicLong insertedAt = new AtomicLong(-1);

        Thread producer = new Thread(() -> {
            try {
                buffer.put(2); // should block until buffer has space
                insertedAt.set(System.currentTimeMillis());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        producer.start();
        Thread.sleep(200); // let producer block

        long beforeDrain = System.currentTimeMillis();
        buffer.take(); // drain — gives producer space
        producer.join(1000);

        assertTrue(insertedAt.get() >= beforeDrain,
                "Producer should have unblocked after consumer drained");
    }

    // ---- Concurrency correctness ----

    @Test
    @Timeout(10)
    void multipleProducersConsumers_noItemsLostOrDuplicated() throws InterruptedException {
        int bufferCapacity = 50;
        int itemsPerProducer = 1000;
        int producerCount = 4;
        int consumerCount = 4;
        int totalItems = itemsPerProducer * producerCount;

        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(bufferCapacity);
        Set<Integer> received = Collections.synchronizedSet(new HashSet<>());
        AtomicInteger consumedCount = new AtomicInteger(0);
        CountDownLatch doneLatch = new CountDownLatch(totalItems);

        ExecutorService exec = Executors.newFixedThreadPool(producerCount + consumerCount);

        // Consumers — run until they've received all expected items
        for (int c = 0; c < consumerCount; c++) {
            exec.submit(() -> {
                try {
                    while (consumedCount.get() < totalItems) {
                        Integer item = buffer.take();
                        if (item == null) break; // poison pill
                        received.add(item);
                        consumedCount.incrementAndGet();
                        doneLatch.countDown();
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
        }

        // Producers — each produces a distinct range of integers
        for (int p = 0; p < producerCount; p++) {
            final int base = p * itemsPerProducer;
            exec.submit(() -> {
                for (int i = 0; i < itemsPerProducer; i++) {
                    try {
                        buffer.put(base + i);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                }
            });
        }

        boolean completed = doneLatch.await(8, TimeUnit.SECONDS);
        exec.shutdownNow();

        assertTrue(completed, "All items should be consumed within timeout");
        assertEquals(totalItems, received.size(),
                "No items should be duplicated — set size must equal total produced");
    }

    // ---- Shutdown correctness ----

    @Test
    @Timeout(3)
    void shutdown_unblocksWaitingConsumers() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(10);
        // buffer is empty — consumer will block on take()

        AtomicInteger nullsReceived = new AtomicInteger(0);
        int consumerCount = 3;
        CountDownLatch exitLatch = new CountDownLatch(consumerCount);

        for (int i = 0; i < consumerCount; i++) {
            new Thread(() -> {
                try {
                    Integer item = buffer.take();
                    if (item == null) nullsReceived.incrementAndGet();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    exitLatch.countDown();
                }
            }).start();
        }

        Thread.sleep(100); // let consumers block
        buffer.shutdown(consumerCount); // insert 3 poison pills

        boolean allExited = exitLatch.await(2, TimeUnit.SECONDS);
        assertTrue(allExited, "All consumers should exit after shutdown");
        assertEquals(consumerCount, nullsReceived.get(),
                "Each consumer should receive exactly one poison pill");
    }

    @Test
    @Timeout(3)
    void shutdown_itemsAlreadyInBuffer_consumedBeforeExit() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(10);
        buffer.put(1);
        buffer.put(2);
        buffer.put(3);

        AtomicInteger consumed = new AtomicInteger(0);
        CountDownLatch exitLatch = new CountDownLatch(1);

        Thread consumer = new Thread(() -> {
            try {
                while (true) {
                    Integer item = buffer.take();
                    if (item == null) break;
                    consumed.incrementAndGet();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } finally {
                exitLatch.countDown();
            }
        });
        consumer.start();

        Thread.sleep(50); // let consumer drain
        buffer.shutdown(1);

        exitLatch.await(2, TimeUnit.SECONDS);
        assertEquals(3, consumed.get(), "All 3 items should be consumed before consumer exits");
    }

    @Test
    void putAfterShutdown_returnsFalse() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(10);
        buffer.shutdown(0);
        assertFalse(buffer.put(1), "put() should return false after shutdown");
    }

    // ---- Observability ----

    @Test
    void sizeAndRemainingCapacity_reflectState() throws InterruptedException {
        BoundedBuffer<Integer> buffer = new BoundedBuffer<>(5);
        assertEquals(0, buffer.size());
        assertEquals(5, buffer.remainingCapacity());

        buffer.put(1);
        buffer.put(2);
        assertEquals(2, buffer.size());
        assertEquals(3, buffer.remainingCapacity());

        buffer.take();
        assertEquals(1, buffer.size());
        assertEquals(4, buffer.remainingCapacity());
    }
}

File: implementation/01 Fundamentals/Concurreny Basics/Producer-Consumer/go/bounded_buffer_test.go
package producerconsumer

import (
	"context"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

// ---- Basic correctness ----

func TestSingleProducerConsumer_OrderPreserved(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](10)
	ctx := context.Background()

	for i := 0; i < 5; i++ {
		if err := buf.Put(ctx, i); err != nil {
			t.Fatalf("Put failed: %v", err)
		}
	}

	for i := 0; i < 5; i++ {
		got, err := buf.Take(ctx)
		if err != nil {
			t.Fatalf("Take failed: %v", err)
		}
		if got != i {
			t.Errorf("expected %d, got %d", i, got)
		}
	}
}

func TestNewBoundedBuffer_InvalidCapacity(t *testing.T) {
	_, err := NewBoundedBuffer[int](0)
	if err == nil {
		t.Error("expected error for capacity 0")
	}
	_, err = NewBoundedBuffer[int](-5)
	if err == nil {
		t.Error("expected error for negative capacity")
	}
}

// ---- Backpressure ----

func TestPut_BlocksWhenFull(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](2)

	ctx := context.Background()
	_ = buf.Put(ctx, 1)
	_ = buf.Put(ctx, 2) // buffer now full

	// Put with a short-lived context — should fail quickly, not block forever
	ctxTimeout, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	err := buf.Put(ctxTimeout, 3)
	if err == nil {
		t.Error("expected Put to fail when buffer is full and context times out")
	}
}

func TestTake_BlocksWhenEmpty(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](5)
	// buffer is empty

	ctxTimeout, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	_, err := buf.Take(ctxTimeout)
	if err == nil {
		t.Error("expected Take to fail when buffer is empty and context times out")
	}
}

func TestProducer_UnblocksAfterConsumerDrains(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](1)
	ctx := context.Background()

	_ = buf.Put(ctx, 1) // fill buffer

	unblocked := make(chan struct{})
	go func() {
		_ = buf.Put(ctx, 2) // should block until drained
		close(unblocked)
	}()

	time.Sleep(50 * time.Millisecond) // let goroutine block

	_, _ = buf.Take(ctx) // drain — unblocks producer

	select {
	case <-unblocked:
		// success
	case <-time.After(1 * time.Second):
		t.Error("producer did not unblock after consumer drained")
	}
}

// ---- Concurrency correctness ----

func TestConcurrentProducersConsumers_NoItemsLost(t *testing.T) {
	const (
		capacity      = 50
		producerCount = 4
		consumerCount = 4
		itemsPerProd  = 1000
		totalItems    = producerCount * itemsPerProd
	)

	buf, _ := NewBoundedBufferReady[int](capacity)

	// Use a cancellable context — cancelled after all producers finish,
	// which unblocks any consumers waiting on an empty buffer.
	ctx, cancel := context.WithCancel(context.Background())

	var consumed int64
	var producerWg sync.WaitGroup
	var consumerWg sync.WaitGroup

	// Start consumers — exit when ctx is cancelled (producers done + buffer drained)
	for c := 0; c < consumerCount; c++ {
		consumerWg.Add(1)
		go func() {
			defer consumerWg.Done()
			for {
				_, err := buf.Take(ctx)
				if err != nil {
					return // context cancelled — exit cleanly
				}
				atomic.AddInt64(&consumed, 1)
			}
		}()
	}

	// Start producers — each writes a distinct range
	for p := 0; p < producerCount; p++ {
		base := p * itemsPerProd
		producerWg.Add(1)
		go func() {
			defer producerWg.Done()
			for i := 0; i < itemsPerProd; i++ {
				if err := buf.Put(ctx, base+i); err != nil {
					return
				}
			}
		}()
	}

	// Wait for all producers to finish, then drain remaining items, then cancel
	go func() {
		producerWg.Wait()
		// Drain: wait until buffer is empty before cancelling consumers
		for buf.Size() > 0 {
			time.Sleep(time.Millisecond)
		}
		// Brief settle — let consumers finish the last in-flight Take
		time.Sleep(10 * time.Millisecond)
		cancel()
	}()

	done := make(chan struct{})
	go func() {
		consumerWg.Wait()
		close(done)
	}()

	select {
	case <-done:
	case <-time.After(8 * time.Second):
		t.Fatalf("test timed out — consumed=%d expected=%d", atomic.LoadInt64(&consumed), totalItems)
	}

	if atomic.LoadInt64(&consumed) != totalItems {
		t.Errorf("items lost: consumed=%d expected=%d", consumed, totalItems)
	}
}

// ---- Shutdown via context cancellation ----

func TestContextCancellation_UnblocksWaitingConsumers(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](10)
	// buffer is empty — consumers will block on Take

	ctx, cancel := context.WithCancel(context.Background())
	consumerCount := 3
	var exitCount int64
	var wg sync.WaitGroup

	for i := 0; i < consumerCount; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			_, err := buf.Take(ctx)
			if err != nil {
				atomic.AddInt64(&exitCount, 1)
			}
		}()
	}

	time.Sleep(50 * time.Millisecond) // let consumers block
	cancel()                          // cancel context — unblocks all consumers

	done := make(chan struct{})
	go func() {
		wg.Wait()
		close(done)
	}()

	select {
	case <-done:
	case <-time.After(1 * time.Second):
		t.Error("consumers did not unblock after context cancellation")
	}

	if atomic.LoadInt64(&exitCount) != int64(consumerCount) {
		t.Errorf("expected %d consumers to exit, got %d", consumerCount, exitCount)
	}
}

// ---- Observability ----

func TestSizeAndRemainingCapacity(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](5)
	ctx := context.Background()

	if buf.Size() != 0 {
		t.Errorf("expected size 0, got %d", buf.Size())
	}
	if buf.RemainingCapacity() != 5 {
		t.Errorf("expected remaining 5, got %d", buf.RemainingCapacity())
	}

	_ = buf.Put(ctx, 1)
	_ = buf.Put(ctx, 2)

	if buf.Size() != 2 {
		t.Errorf("expected size 2, got %d", buf.Size())
	}
	if buf.RemainingCapacity() != 3 {
		t.Errorf("expected remaining 3, got %d", buf.RemainingCapacity())
	}

	_, _ = buf.Take(ctx)

	if buf.Size() != 1 {
		t.Errorf("expected size 1, got %d", buf.Size())
	}
}

// ---- Race detector validation ----
// Run with: go test -race ./...

func TestRaceDetector_ConcurrentAccess(t *testing.T) {
	buf, _ := NewBoundedBufferReady[int](20)
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	var wg sync.WaitGroup

	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			for j := 0; j < 100; j++ {
				_ = buf.Put(ctx, id*100+j)
			}
		}(i)
	}

	for i := 0; i < 5; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < 100; j++ {
				_, _ = buf.Take(ctx)
			}
		}()
	}

	wg.Wait()
}